modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientInboundMessageHandler.java

```diff
@@ -0,0 +1,484 @@
+import org.apache.ignite.internal.schema.SchemaDescriptor;
+import org.apache.ignite.internal.table.IgniteTablesInternal;
+import org.apache.ignite.internal.table.TableImpl;
+import org.apache.ignite.internal.table.TupleBuilderImpl;
+import org.apache.ignite.lang.IgniteException;
+import org.apache.ignite.table.Table;
+import org.apache.ignite.table.Tuple;
+import org.msgpack.core.MessageFormat;
+import org.msgpack.core.buffer.ByteBufferInput;
+import org.slf4j.Logger;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.nio.ByteBuffer;
+import java.util.BitSet;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Handles messages from thin clients.
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class ClientInboundMessageHandler extends ChannelInboundHandlerAdapter {
+    /** Logger. */
+    private final Logger log;
+
+    /** API entry point. */
+    private final Ignite ignite;
+
+    /** Context. */
+    private ClientContext clientContext;
+
+    /**
+     * Constructor.
+     *
+     * @param ignite Ignite API entry point.
+     * @param log Logger.
+     */
+    public ClientInboundMessageHandler(Ignite ignite, Logger log) {
+        assert ignite != null;
+        assert log != null;
+
+        this.ignite = ignite;
+        this.log = log;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void channelRead(ChannelHandlerContext ctx, Object msg) throws IOException {
+        var buf = (ByteBuffer) msg;
+
+        var unpacker = getUnpacker(buf);
+        var packer = getPacker();
+
+        if (clientContext == null)
+            handshake(ctx, unpacker, packer);
+        else
+            processOperation(ctx, unpacker, packer);
+    }
+
+    private void handshake(ChannelHandlerContext ctx, ClientMessageUnpacker unpacker, ClientMessagePacker packer)
+            throws IOException {
+        try {
+            var clientVer = ProtocolVersion.unpack(unpacker);
+
+            if (!clientVer.equals(ProtocolVersion.LATEST_VER))
+                throw new IgniteException("Unsupported version: " +
+                        clientVer.major() + "." + clientVer.minor() + "." + clientVer.patch());
+
+            var clientCode = unpacker.unpackInt();
+            var featuresLen = unpacker.unpackBinaryHeader();
+            var features = BitSet.valueOf(unpacker.readPayload(featuresLen));
+
+            clientContext = new ClientContext(clientVer, clientCode, features);
+
+            log.debug("Handshake: " + clientContext);
+
+            var extensionsLen = unpacker.unpackMapHeader();
+            unpacker.skipValue(extensionsLen);
+
+            // Response.
+            ProtocolVersion.LATEST_VER.pack(packer);
+
+            packer.packInt(ClientErrorCode.SUCCESS)
+                    .packBinaryHeader(0) // Features.
+                    .packMapHeader(0); // Extensions.
+
+            write(packer, ctx);
+        } catch (Throwable t) {
```

Comment:
}
    catch (Exception e) {

Reply:
Fixed (can we use checkstyle to catch this?)